nextcloud/server
# Add a mechanism to register crash reporters

This adds a crash reporter registry, which is meant for third party apps
to hook into the error logging/reporting chain. This way, external tools
like Sentry can be used to track and aggregate system crashes.

Out of personal interest (I'm using Sentry to monitor some Laravel apps), I looked into integrating Sentry into Nextcloud. As a first attempt, I just used the provided SDK and registered error/exception handlers. While this works for some triggered errors, it does not report and exception as we catch and handle them in places like `index.php`. Thus, I needed to find a way to hook into the exception reporting chain and found out that our logger is usually called to log the exception. Therefore, I've added a little mechanism that allows apps to register any reporter that implements a certain interface. If no reporters are registered, the registry doesn't do anything.

Usually, I would have opened a ticket first to discuss the general idea but since it's a rainy Sunday afternoon  and I was motivated to invest some time on this, I started the implementation right away and it was finished quickly.

Let me know what you think. The world doesn't end if anyone feels like this is too specific and we shouldn't add this to our public API. After all, someone has to maintain it. Still, I think it's a nice little enhancement and used correctly, the usage of Sentry will help detect more unnoticed errors that usually get lost in log files.

~~If we decide to integrate this I have to add some tests.~~ done

Steps to test
* Create an account on sentry.io or set up Sentry via Docker
* https://github.com/ChristophWurst/nextcloud_sentry
* Add the SDN via `php occ config:system:set sentry.dsn --value=<DSN>`
* Let something fail
* See the report on Sentry

cc @LukasReschke @nickvergessen @rullzer
